### PR TITLE
Lima cgroupsv2

### PR DESCRIPTION
**Pull request by Devatoria**
## What does this PR do?

- [ ] Adds new functionality
- [x] Alters existing functionality
- [ ] Fixes a bug
- [ ] Improves documentation or testing

Please briefly describe your changes as well as the motivation behind them:

We are planning for a cgroups v2 support soon and need to adapt the local env for testing and debugging. Colima was not allowing an easy switch to cgroups v2. Lima, used by Colima, allows us to custom some more settings. Because of this, here's what the PR changes:
- It moves from Colima to Lima for the local env
  - It keeps cgroups v1 on Alpine as the default behavior when starting the stack locally
  - It creates a local stack based on Ubuntu for cgroups v2 (no hybrid)
- It updates the `Makefile` as well as the contributing documentation

All functionalities of the local stack should be working on both cgroups v1 and v2 (things such as longhorn for instance). The chaos-controller runs fine but any disruption relying on cgroups will fail with the cgroups v2.

### How to switch to cgroups v2?

`CGROUPS=v2 make lima-all` will create the stack with Ubuntu and cgroups v2. You must delete and recreate the stack to switch between the cgroup mode (`make lima-stop`).

## Code Quality Checklist

- [ ] The documentation is up to date.
- [ ] My code is sufficiently commented and passes continuous integration checks.
- [ ] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md)).

## Testing

- [ ] I leveraged continuous integration testing
    - [ ] by depending on existing `unit` tests or `end-to-end` tests.
    - [ ] by adding new `unit` tests or `end-to-end` tests.
- [ ] I manually tested the following steps:
    - `x`
    - [ ] locally.
    - [ ] as a canary deployment to a cluster.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -69,116 +58,131 @@ handler:
...
-# Generate code
+## Generate code
 generate: controller-gen
 	$(CONTROLLER_GEN) object:headerFile=./hack/boilerplate.go.txt paths="./..."
 
-colima-build-manager: manager
-	nerdctl build --namespace ${CONTAINERD_REGISTRY_PREFIX} --build-arg TARGETARCH=${OS_ARCH} -t ${MANAGER_IMAGE} -f bin/manager/Dockerfile ./bin/manager/
+# Lima actions
+## Create a new lima cluster and deploy the chaos-controller into it
+lima-all: lima-start lima-install-cert-manager lima-build-all lima-install
```
Comment (ptnapoleon) [inline Makefile:105]: 😭 this is so useful

**Review by ptnapoleon — APPROVED** [review]
I had no issues starting alpine or ubuntu nodes with lima

**Review by taihuynh167 — APPROVED** [review]
🥳 I added some question for my knowledge

File: Makefile [hunk 2]
```diff
@@ -69,116 +58,131 @@ handler:
...
+## Stop and delete the lima cluster
+lima-stop:
+	limactl stop -f ${LIMA_PROFILE}
+	limactl delete ${LIMA_PROFILE}
+	$(MAKE) lima-clean
+
+## Start the lima cluster, pre-cleaning kubectl config
+lima-start: lima-clean
+	limactl start --tty=false --name=${LIMA_PROFILE} ./${LIMA_CONFIG}.yaml
+	limactl shell ${LIMA_PROFILE} sudo sed 's/default/lima/g' /etc/rancher/k3s/k3s.yaml >> ~/.kube/config
```
[the remark this answers is not in the document]
Comment (luphaz) [reply Makefile:177]: When I used `kubens` to switch to `chaos-engineering` this is wiping out the `~/.kube/config` file. Keeping only lima context 😬.

[the remark this answers is not in the document]
Comment (luphaz) [reply Makefile:177]: Quick search we may be able to do something like:
* https://medium.com/@jacobtomlinson/how-to-merge-kubernetes-kubectl-config-files-737b61bd517d

It seems to work as expected

```suggestion
lima-kubectx: lima-clean
	limactl shell ${LIMA_PROFILE} sudo sed 's/default/lima/g' /etc/rancher/k3s/k3s.yaml > ~/.kube/config_lima
	KUBECONFIG=~/.kube/config:~/.kube/config_lima kubectl config view --flatten > /tmp/config
	mv /tmp/config ~/.kube/config
	rm ~/.kube/config_lima
	kubectx ${LIMA_PROFILE}

lima-start:
	limactl start --tty=false --name=${LIMA_PROFILE} ./${LIMA_CONFIG}.yaml
	$(MAKE) lima-kubectx
```

**Review by luphaz — APPROVED** [review]

File: Makefile [hunk 3]
```diff
@@ -69,116 +58,131 @@ handler:
...
-# Build the docker images
-# to ease minikube/colima base minikube docker image load, we are using tarball
-# we encountered issue while using docker+containerd on colima to load images from docker to containerd
-# this Make target are used by the CI
-minikube-build-manager: manager
+docker-build-manager: manager
 	docker build --build-arg TARGETARCH=${OS_ARCH} -t ${MANAGER_IMAGE} -f bin/manager/Dockerfile ./bin/manager/
-	docker save -o ./bin/manager/manager.tar ${MANAGER_IMAGE}
-	minikube image load --daemon=false --overwrite=true ./bin/manager/manager.tar
+	docker save ${MANAGER_IMAGE} -o ./bin/manager.tar.gz
```
Comment (luphaz) [inline Makefile:136]: You may want to put those files under each dedicated bins or add them to the `.gitignore`
```suggestion
	docker save ${MANAGER_IMAGE} -o ./bin/manager/manager.tar.gz
```

**Review by aymericDD — APPROVED** [review]

File: examples/level_node.yaml [hunk 4]
```diff
@@ -11,6 +11,6 @@ metadata:
 spec:
   level: node # impact the whole node instead of a single pod
   selector:
-    kubernetes.io/hostname: colima # when node level is enabled, the selector applies to nodes and not pods
+    kubernetes.io/hostname: lima-lima # when node level is enabled, the selector applies to nodes and not pods
```
[the remark this answers is not in the document]
Comment (ptnapoleon) [reply examples/level_node.yaml:14]: `lima` prefixes the host name with `lima-`

File: Makefile [hunk 5]
```diff
@@ -69,116 +58,131 @@ handler:
...
+## Build and import the handler image in lima
+lima-build-handler: docker-build-handler
+	limactl copy ./bin/handler.tar.gz ${LIMA_PROFILE}:/tmp/
+	limactl shell ${LIMA_PROFILE} -- sudo k3s ctr i import /tmp/handler.tar.gz
+
+## Remove lima references from kubectl config
+lima-clean:
+	kubectl config delete-cluster ${LIMA_PROFILE} || true
+	kubectl config delete-context ${LIMA_PROFILE} || true
+	kubectl config delete-user ${LIMA_PROFILE} || true
```
[the remark this answers is not in the document]
Comment (ptnapoleon) [reply Makefile:165]: TIL

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (ptnapoleon) [reply Makefile:177]: > When I used kubens to switch to chaos-engineering this is wiping out the ~/.kube/config file. Keeping only lima context 😬.

Oh no, I didn't realize this happened to me too

File: lima_ubuntu.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+vmType: "vz"
+
+# enable rosetta for x86_64 arch compatibility
+rosetta:
+  enabled: true
+  binfmt: true
+
+images:
+- location: "https://cloud-images.ubuntu.com/releases/22.04/release/ubuntu-22.04-server-cloudimg-amd64.img"
+  arch: "x86_64"
```
[the remark this answers is not in the document]
Comment (Devatoria) [reply lima_ubuntu.yaml]: This file is a copy of the original file given by lima, and it doesn't contain it. I guess we can add it if needed, not sure if it's useful.

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (Devatoria) [reply Makefile]: TIL too

File: Makefile (hunk 3 above)
Comment (Devatoria) [reply Makefile]: Good catch, done.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (Devatoria) [reply Makefile]: We can do that, and this is what is suggested in the lima examples, however it means that it'll only be working for the terminal session which ran the make command as `KUBECONFIG` will be only populated there.

[the remark this answers is not in the document]
Comment (luphaz) [reply Makefile]: The commands I provided combines several `kubeConfig` into a single one, this imply all other terminal session will still benefits from the "merged" contexts.

[the remark this answers is not in the document]
Comment (Devatoria) [reply Makefile]: Oh damn I missed the `kubectl config view` part with the redirect 🤦 sorry.

**Review by aymericDD — APPROVED** [review]

File: Makefile [hunk 7]
```diff
@@ -197,16 +208,19 @@ else
...
 
-colima-stop:
-	colima stop
+# CI-specific actions
+
+## Minikube builds for e2e tests
+minikube-build-all: minikube-build-manager minikube-build-injector minikube-build-handler
+
+minikube-build-manager: docker-build-manager
+	minikube image load --daemon=false --overwrite=true ./bin/manager.tar.gz
```
Comment (luphaz) [inline Makefile]:
```suggestion
	minikube image load --daemon=false --overwrite=true ./bin/manager/manager.tar.gz
```

File: Makefile [hunk 8]
```diff
@@ -197,16 +208,19 @@ else
...
 
-colima-start:
-	colima start --runtime containerd --kubernetes --kubernetes-version=${KUBERNETES_VERSION}+k3s1 --cpu 4 --memory 4 --disk 50
-# We also install iproute2-tc that contains 'tc' and enables to perform debugging from the VM
-	colima ssh -- sudo apk add iproute2-tc
+minikube-build-injector: docker-build-injector
+	minikube image load --daemon=false --overwrite=true ./bin/injector.tar.gz
```
Comment (luphaz) [inline Makefile]:
```suggestion
	minikube image load --daemon=false --overwrite=true ./bin/injector/injector.tar.gz
```

File: Makefile [hunk 9]
```diff
@@ -197,16 +208,19 @@ else
...
 
-colima-install-nerdctl:
-	sudo colima nerdctl install
+minikube-build-handler: docker-build-handler
+	minikube image load --daemon=false --overwrite=true ./bin/handler.tar.gz
```
Comment (luphaz) [inline Makefile]:
```suggestion
	minikube image load --daemon=false --overwrite=true ./bin/handler/handler.tar.gz
```

**Pull request merged**